### PR TITLE
Expose ECDH functions using EVP_derive_*

### PR DESCRIPTION
**Pull request by ueno**
These functions are required to implement ECDH key agreement in TLS in
a FIPS compliant manner, i.e., remote public key validation according
to SP800-56A is performed.

Fixes: #6
Signed-off-by: Daiki Ueno <dueno@redhat.com>

### CODE REVIEW

**Comment by ueno** [comment]
The API (`SharedKeyECDH`) is currently modelled after TLS [usage](https://github.com/golang/go/blob/dev.boringcrypto.go1.18/src/crypto/tls/key_schedule.go#L109). @derekparker do you have any opinion on that?

**Comment by simo5** [comment]
I see so they do not exposed a full ECDH, sounds reasonable to expose only what's used.

**Review by derekparker — APPROVED** [review]
LGTM

**Comment by derekparker** [comment]
+1'ed, and @ueno your implementation is perfect, let's expose only what we have to. Thanks for the patch!

**Comment by derekparker** [comment]
@dbenoit17 can you take a look and +2 this PR before we merge?

**Review by simo5 — APPROVED** [review]
LGTM

File: openssl/goopenssl.h
Comment (ueno) [reply openssl/goopenssl.h:793]: The upstream Go interface (in `crypto/elliptic`) only exposes raw primitives, e.g., scalar multiplication. Afaik those EVP functions are for X9.63 KDF, but there is no use of it in `crypto/tls` either (because in TLS we combine raw ECDH with HKDF to derive key materials).

**Comment by ueno** [comment]
I'm merging this after the discussion yesterday (it already has 2 approvals).

**Pull request merged**
